nm ipsec: Do not set default values

cathay4t opened the pull request:
Introducing new ipsec option `nm-auto-defaults: true|false`, default to
true. When set to false, instruct NetworkManager to not set default
values but pass only desired values to ipsec libreswan daemon as it is.

With this, host-to-host(P2P) mode is not required to set `leftmodecfgclient: no`
explicitly which align with `ipsec.conf` manpage of libreswan.

It is recommend to set `nm-auto-defaults: false` and perform following
steps to align your YAML with libreswan daemon default values:
 * Upgrade NetworkManager-libreswan to 1.2.28 or later version.
 * Set `leftmodecfgclient: yes` explicitly if you need pull config from
   remote.
 * Add `@` to `leftid` and `rightid` if you don't want it been resolved
   to IP address. (e.g. `leftid: '@cli-a.example.org'`).
 * Define `rightsubnet` explicitly. (e.g.`rightsubnet: 0.0.0.0/0`)

Resolves: https://issues.redhat.com/browse/RHEL-26350

cathay4t commented:
The CI failure is caused by CentOS stream 9 not having NetworkManager-libreswan 1.2.28.

ihuguet commented:
The more I think about this strategy, the less I like it. Applying a different config when retrying is not a good idea, IMO. If the first attempt fails for a different reason other than the NM-libreswan version (which happens often), we'll apply a configuration different to what the user intended.

In order to support old NetworkManager-libreswan version we should better check the plugin version. In order to support old yaml, maybe we should add a `nm-auto-defaults` field. I know that you don't like that, and I neither like it, but as you always say, better explicit than implicit.

Or maybe it would be better just to apply the change as a breaking change and not supporting old plugin and yamls. We can do it in RHEL 10 and keep the old behavior in RHEL 9.

cathay4t commented:
Yeah. I think introduce `nm-auto-defaults: true` is the only way to preserve full backwards compatibility. Let me make amends.

ihuguet commented:
It was a good attempt, btw, trying to make it to work seamlessly is the first thing that we must try...

Regarding `nm-auto-defaults`, maybe we can document somehow that it is a transition field and in the future it won't be needed, and people should move their yamls to `nm-auto-defaults: false` mode?

cathay4t commented:
> It was a good attempt, btw, trying to make it to work seamlessly is the first thing that we must try...
> 
> Regarding `nm-auto-defaults`, maybe we can document somehow that it is a transition field and in the future it won't be needed, and people should move their yamls to `nm-auto-defaults: false` mode?

Considering OpenShift does not care whether OS is RHEL 9 or RHEL 10, their users are expecting the YAML works forever.
I think the correct thing we can do is add `nm-auto-default: false` to all the documents of nmstate ipsec.

cathay4t committed:
ipsec: New option `nm-auto-defaults: false`

Introducing new ipsec option `nm-auto-defaults: true|false`, default to
true. When set to false, instruct NetworkManager to not set default
values but pass only desired values to ipsec libreswan daemon as it is.

With this, host-to-host(P2P) mode is not required to set `leftmodecfgclient: no`
explicitly which align with `ipsec.conf` manpage of libreswan.

It is recommend to set `nm-auto-defaults: false` and perform following
steps to align your YAML with libreswan daemon default values:
 * Upgrade NetworkManager-libreswan to 1.2.28 or later version.
 * Set `leftmodecfgclient: yes` explicitly if you need pull config from
   remote.
 * Add `@` to `leftid` and `rightid` if you don't want it been resolved
   to IP address. (e.g. `leftid: '@cli-a.example.org'`).
 * Define `rightsubnet` explicitly. (e.g.`rightsubnet: 0.0.0.0/0`)

Resolves: https://issues.redhat.com/browse/RHEL-26350

Signed-off-by: Gris Ge <fge@redhat.com>
